src/event_emitter/base.cr

```diff
@@ -1,54 +1,67 @@
 module EventEmitter
-  class Base 
-    @all_handlers = [] of Proc(String, Any, Nil)
-    @channels = Hash(String, Array(Channel::Unbuffered(Any))).new
-
-    def all(&block : String, Any ->)
-      @all_handlers << ->(event : String, response : Any) {
-        block.call(event, response)
-      }
+  class Base
+    alias Event = NamedTuple(event: String | Regex, listener: Any ->, once: Bool, id: Int32)
+    @events = [] of Event
+
+    # Listen to all events and run the specified block. This is the
+    # same as doing `emitter.on(/.*/) { ... }`
+    def all(params = {} of Any => Any, &block : Any ->)
+      on(/.*/, &block)
     end
 
-    def on(event, &block : Any ->)
-      event = event.to_s
-      channel = Channel::Unbuffered(Any).new
+    # Listen for a particular `event` and fire the block
+    # every time that event is called. If `once` is `true`
+    # the event will be removed after it is fired the
+    # first time.
+    def on(event, *, once = false, &block : Any ->)
+      event = event.to_s unless event.is_a?(Regex)
 
-      if @channels.has_key? event
-        @channels[event] << channel
-      else
-        @channels[event] = [channel]
-      end
-
-      spawn do
-        loop do
-          block.call(channel.receive)
-        end
-      end
+      id = @events.empty? ? 0 : @events.last[:id]
+      @events << {
+        event: event.as(Regex | String),
+        listener: block,
+        once: once,
+        id: id
+      }.as(Event)
+      id
     end
 
+    # Sugar for `on(event, once: true)`. Will fire the
+    # event once and then remove the listener.
     def once(event, &block : Any ->)
+      on(event, once: true, &block)
+    end
+
+    # Emit an event with the specified `arg`
+    def emit(event : Symbol | String | Regex, arg = nil)
       event = event.to_s unless event.is_a?(String)
-      channel = Channel::Unbuffered(Any).new
-      if @channels.has_key? event
-        @channels[event] << channel
-      else
-        @channels[event] = [channel]
-      end
+      arg = EventEmitter.any(arg)
+
+      @events.each do |e|
+        case e[:event]
+        when Regex
+          if event =~ e[:event]
+            listener = e[:listener]
+            listener.call(arg)
```

Comment:
This `#call` makes all events synchronous, previously they were called inside another fiber with `spawn`

Reply:
This is the one thing I've had a hard time resolving. I'll keep thinking on how to make things async.

Reply:
Ok I pushed a change that wraps that whole thing in a spawn and then calls `Fiber.yield` before the end of the method. I don't know that it's actually doing anything useful, but I'm at the end of my rope with async stuff.